### DISCUSSION TITLE
omada SDN integration?

### DISCUSSION THREAD

**Discussion by FlyingToto** [netalertx/NetAlertX#707] [Ideas]
Good afternoon,
Any "easy" way to integrate with omada SDN? this is an example of data that we can pull from omada:

```
MAC                                   IP                   hostname            switch(port)/SSID(AccessPoint)
50-02-91-01-02-03     192.168.0.153   mickey                     swtichRoom2  (20)
60-DD-8E-01-02-04    192.168.0.226   pluto                        floor2wifi (floorSouth2AP)
FC-AA-14-01-02-05     192.168.0.10    banana                     switchRoom5 (12)
```
omada also has the notion of  "sites" which basically divides networks into multiple locations like and the scan is on a per-site basis.

- First and foremost, thank you for NetAlertX, it is a great piece of software and it should be integrated to everything IMHO.
- I have been using tplink omada SDN to manage most of my devices and while fixing a limitation of the tplink code on my own, I cam across this great python omada library which provides very interesting insight from omada sdn in particular the ability to extract clients by MAC addresses mapped to specific omada Switches ports and Access Points SSID... Is there any way to integrate this data (ideally as a plugin) into NetAlertX? 

1) install python: sudo apt install python3-pip
2) install the omada python api: pip install tplink-omada-client
3) update the parameters CHANGE_ME values in the scipt below
omada_username="CHANGEME_mrpotatoe"
omada_password="CHANGEME_lfdaruiRWGFD335qw324z"
omada_site="CHANGEME_homesweethome"
omada_url="https://CHANGEME_omada.mylocaldomain" 
omada -t myomada target --url $omada_url --user $omada_username --password $omada_password --site $omada_site --set-default
omada clients

**Comment by jokob-sk** [comment]
Hi @FlyingToto ,

I like this idea! I regrettably don't own these devices so I need to rely to someone (like you?😃) to implement it. Can you code? I can prepare a skeleton for you with the files and then leave it to you to fill in the blanks?  I can _try_ to do it myself and you test only, but usually that takes very long as I can't quickly iterate so the development is pretty long and frustrating.

Let me know what you think.

**Reply by FlyingToto** [reply]
well, I can definitely give it a shot...  I know a little bit of shell, and python.... 

- if you give me some api to pull a record and update it back we should be able to make it work.  
- we would need to be able to schedule as well so it gets refreshed periodically.
- also is there a way to include these tplink python module with netalertx?

**Reply by jokob-sk** [reply]
Thanks a lot 🙏 I'll prepare a solid starting point for you and include relevant packages. Most of the above (settings, scheduling, UI, updating of the existing records, creation of new ones) is handled by the plugins system: https://github.com/jokob-sk/NetAlertX/blob/main/docs/PLUGINS_DEV.md

I should be able to have a starting point ready this weekend.

**Reply by FlyingToto** [reply]
yeah I was looking at the plugins doc but I am afraid that might go a bit beyond my coding skills... one thing that I thought of doing was to use the export/import csv feature as a one-time manual udpate thing. 

BTW,  I do have a few questions:


1)  can I use the device PORT field to store the SSID string name ?
-> omada nicely provides the port for each switch that a device is connected to and I netalertX does allow me to track the port number per device. But for Access points, omada provides the SSID of the AP that the device is connected to... 

2) since I can populate the switch/AP name per device and the port/SSID, will that build the network graph automatically for me? 
 (that would be so cool! I better buy the older tplink switches omada compatible on ebay before people learn about that!)


3) I cant remember if someone already asked, but any places to track VLANs in netalertX?   or maybe even to allow custom fields? (for instance omada's api provide more details like PoE  and/or network speed etc...) here is an example:

Port: 47 - Port47
    Type:        1
    Profile:     All - ☐
    Enabled:     ☑
    Status:      ☐
    PoE:         ☑    0.0W
    Link Speed:  SPEED_10_MBPS (Max: SPEED_1_GBPS)
    Transmitted: 0.0MB
    Received:    0.0MB
    Operation:   switching
    Limit Mode:  OFF
Port: 48 - Port48
    Type:        1
    Profile:     All - ☐
    Enabled:     ☑
    Status:      ☑
    PoE:         ☑    0.0W
    Link Speed:  SPEED_1_GBPS (Max: SPEED_1_GBPS)
    Transmitted: 54.4GB
    Received:    112.8GB
    Operation:   switching
    Limit Mode:  OFF
Port: 49 - Port49
    Type:        3
    Profile:     All - ☐
    Enabled:     ☑
    Status:      ☐
    PoE:          Not supported
    Link Speed:  SPEED_10_MBPS (Max: SPEED_1_GBPS)
    Transmitted: 0.0MB
    Received:    0.0MB
    Operation:   switching
    Limit Mode:  OFF

**Reply by jokob-sk** [reply]
Hey @FlyingToto, 

Let's continue the discussion here as I created an issue/task for this work: https://github.com/jokob-sk/NetAlertX/issues/708

I answered your questions there as well. 

Thanks in advance for the help!
